src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1381,24 +1381,40 @@ namespace Microsoft.FSharp.Collections
         [<CompiledName("Last")>]
         let last (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                res
-            else
-                invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+            match source with
+            | :? ('T[]) as a -> 
+                if a.Length = 0 then invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                else a.[a.Length - 1]
+            | :? IList<'T> as a -> 
+                if a.Count = 0 then invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                else a.[a.Count - 1]
+            | _ -> 
+                use e = source.GetEnumerator()
+                if e.MoveNext() then
+                    let mutable res = e.Current
+                    while (e.MoveNext()) do res <- e.Current
+                    res
+                else
+                    invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
 
         [<CompiledName("TryLast")>]
         let tryLast (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                Some res
-            else
-                None
+            match source with
+            | :? ('T[]) as a -> 
+                if a.Length = 0 then None
+                else Some(a.[a.Length - 1])
+            | :? IList<'T> as a -> 
```

Comment:
How can `ICollection`   be indexed? It does not have an `.Item` member.

Reply:
@goswinr, you are absolutely right, I just checked for `Count`. And a `Queue` would have to be enumerated anyway, so that would only benefit a `Stack`, but than you have to create too many special cases I'd reckon.